### DISCUSSION TITLE
gmp/gmp-model

### DISCUSSION THREAD

**Discussion by giscus** [cyub/go-1.14.13#4] [Announcements]
# gmp/gmp-model

  GMP模型
  #

Golang的一大特色就是Goroutine。Goroutine是Golang支持高并发的重要保障。Golang可以创建成千上万个Goroutine来处理任务，将这些Goroutine分配、负载、调度到处理器上采用的是G-M-P模型。

  什么是Goroutine
  #

Goroutine = Golang + Coroutine。Goroutine是golang实现的协程，是用户级线程。Goroutine具有以下特点：

相比线程，其启动的代价很小，以很小栈空间启动（2Kb左右）
能够动态地伸缩栈的大小，最大可以支持到Gb级别
工作在用户态，切换成很小
与线程关系是n:m，即可以在n个系统线程上多工调度m个Goroutine

https://go.cyub.vip/gmp/gmp-model.html

**Comment by Austinxian** [comment]
总结的第三点，应该是先从global queue获取可执行的G，然后再从其他P那里获取吧

**Reply by cyub** [reply]
是的。上面总结有误。具体源码可见：https://github.com/golang/go/blob/go1.14.13/src/runtime/proc.go#L2156-L2252

**Reply by shark2shark** [reply]
有点不明白，global queue不是有锁吗，有锁的开销不是更大吗，为什么会优先从global queue获取呢

**Reply by yezige** [reply]
> 上面生命周期流程说明：
> - runtime 创建最初的线程 m0 和 goroutine g0，并把两者进行关联（g0.m = m0)
> - 调度器初始化：设置M最大数量，P个数，栈和内存出事，以及创建 GOMAXPROCS个P，示例代码中的 main 函数是 main.main，runtime 中也有 1 个 > main 函数 ——runtime.main，代码经过编译后，runtime.main 会调用 main.main，程序启动时会为 runtime.main 创建 goroutine，称它为 main goroutine 吧，然后把 main goroutine 加入到 P 的本地队列。
> - 启动 m0，m0 已经绑定了 P，会从 P 的本地队列获取 G，获取到 main goroutine。
> - G 拥有栈，M 根据 G 中的栈信息和调度信息设置运行环境
> - M 运行 G
> - G 退出，再次回到 M 获取可运行的 G，这样重复下去，直到 main.main 退出，runtime.main 执行 Defer 和 Panic 处理，或调用 runtime.exit 退出程序。

@cyub 生命周期那里最后一步，G退出，这个G应该是m0获取到的main goroutine，G退出应该就代表runtime.main已经退出了，也就没有`再次回到 M 获取可运行的 G，直到 main.main 退出` 之说了吧，不知道我的理解对不对

**Comment by moqsien** [comment]
g0负责调度其他G，那么g0是怎么在线程上调度的呢，这个好像没讲清楚。

**Reply by yaoguangduan** [reply]
g0是从rt0_xx用汇编启动的

**Comment by richard667** [comment]
Golang的线程模型的图，M应该在内核空间？

**Comment by y1yang0** [comment]
博主你好，请问能否在微信公众号引用你文章中GMP模型的图片，我们会标明出处

谢谢

**Reply by cyub** [reply]
可以
